hydroflow_lang/src/graph/ops/join_fused.rs

```diff
@@ -0,0 +1,333 @@
+use proc_macro2::{Ident, TokenStream};
+use quote::{quote_spanned, ToTokens};
+use syn::spanned::Spanned;
+use syn::{parse_quote, Expr, ExprCall};
+
+use super::{
+    DelayType, FlowProperties, FlowPropertyVal, OperatorCategory, OperatorConstraints,
+    OperatorWriteOutput, Persistence, WriteContextArgs, RANGE_0, RANGE_1,
+};
+use crate::diagnostic::{Diagnostic, Level};
+use crate::graph::{OpInstGenerics, OperatorInstance};
+
+/// > 2 input streams of type <(K, V1)> and <(K, V2)>, 1 output stream of type <(K, (V1, V2))>
+///
+/// `join_fused` takes two arguments, they are the left hand side and right hand side reducing/folding operation respectively.
+/// The syntax is a bit unique for them, but it is similar to specifying an enum in rust. For example:
+/// ```hydroflow,ignore
+/// join_fused(Fold(|| "default value", |x, y| FoldingFunction(x, y)), Reduce(|x, y| ReducingFunction(x, y)))
+/// join_fused(FoldFrom(|x| "conversion function", |x, y| FoldingFunction(x, y)), Reduce(|x, y| ReducingFunction(x, y)))
+/// ```
+/// The three currently supported fused operator types are `Fold(Fn() -> A, Fn(A, T) -> A)`, `Reduce(Fn(A, A) -> A)`, and `FoldFrom(Fn(T) -> A, Fn(A, T) -> A)`
+///
+/// `join_fused` first performs a fold_keyed/fold_reduced operation on each input stream before performing joining. See `join()`. There is currently no equivalent for `FoldFrom` in hydroflow operators.
+///
+/// For example, the following two hydroflow programs are equivalent, the former would optimize into the latter:
+///
+/// ```hydroflow
+/// source_iter(vec![("key", 0), ("key", 1), ("key", 2)])
+///     -> reduce_keyed(|x: &mut _, y| *x += y)
+///     -> [0]my_join;
+/// source_iter(vec![("key", 2), ("key", 3)])
+///     -> fold_keyed(|| 1, |x: &mut _, y| *x *= y)
+///     -> [1]my_join;
+/// my_join = join_multiset()
+///     -> assert_eq([("key", (3, 6))]);
+/// ```
+///
+/// ```hydroflow
+/// source_iter(vec![("key", 0), ("key", 1), ("key", 2)])
+///     -> [0]my_join;
+/// source_iter(vec![("key", 2), ("key", 3)])
+///     -> [1]my_join;
+/// my_join = join_fused(Reduce(|x, y| *x += y), Fold(|| 1, |x, y| *x *= y))
```

Comment:
This was the first point I started to understand that you choose one of `Reduce` or `Fold` per input, not both.

Reply:
I think the new additions should make it clearer, earlier now.